diff --git a/native/src/seal/util/polyarithsmallmod.h b/native/src/seal/util/polyarithsmallmod.h
index d0fef603..211ae629 100644
--- a/native/src/seal/util/polyarithsmallmod.h
+++ b/native/src/seal/util/polyarithsmallmod.h
@@ -739,10 +739,21 @@ namespace seal
             }
 #endif
             auto poly_modulus_degree = result.poly_modulus_degree();
+            if (mono_exponent > 2 * poly_modulus_degree) {
+                mono_exponent %= 2 * poly_modulus_degree;
+            }
+            bool neg = false;
+            if (mono_exponent > poly_modulus_degree) {
+                mono_exponent -= poly_modulus_degree;
+                neg = true;
+            }
             SEAL_ITERATE(iter(poly, mono_coeff, modulus, result), coeff_modulus_size, [&](auto I) {
                 negacyclic_multiply_poly_mono_coeffmod(
                     get<0>(I), poly_modulus_degree, get<1>(I), mono_exponent, get<2>(I), get<3>(I), pool);
             });
+            if (neg) {
+                negate_poly_coeffmod(result, coeff_modulus_size, modulus, result);
+            }
         }
 
         inline void negacyclic_multiply_poly_mono_coeffmod(
